test/input_gen/genInput.py

```diff
@@ -198,41 +202,52 @@ def fc_tf(x, kernel, label, bias, activation, train=False, loss='mse', opt='sgd'
 # tested with tf 1.14.0
 # @param[in] x input
 # @param[in] trainable
-# @return bn output, [updated_gamma, updated_beta], grad_result (0. dx / 1. gamma / 2. beta / 3. mean / 4. variance)
+# @return input_variables, bn output, output_variables, grad_result (0. dx / 1. gamma / 2. beta / 3. mean / 4. variance)
 # for updated_gamma, updated_beta, x <- x - grad is used for easier calculation
-def bn_tf(x, trainable=False):
+def bn_tf(x, *, trainable=True, init_beta=gen_tensor, init_gamma=gen_tensor, axis=[1, 2, 3]):
     tf.compat.v1.reset_default_graph()
     tf_input = tf.compat.v1.placeholder(
         dtype=dtypes.float32, shape=x.shape, name='input')
 
     bnlayer = tf.keras.layers.BatchNormalization(
-        axis=0,
+        axis=axis,
         trainable=trainable,
+        momentum=1.0,
         gamma_initializer=gen_tensor,
         beta_initializer=gen_tensor)(tf_input)
 
-    bn_variables = tf.compat.v1.get_collection(tf.GraphKeys.GLOBAL_VARIABLES,
+    bn_variables = tf.compat.v1.get_collection(tf.compat.v1.GraphKeys.GLOBAL_VARIABLES,
                                                scope='batch_normalization')
+
     input_variables = [tf_input] + bn_variables
 
     grad = tf.gradients(bnlayer, input_variables)
 
+    f_dict = {tf_input: x, tf.keras.backend.learning_phase(): trainable}
+
     with tf.compat.v1.Session() as sess:
+      with tf.compat.v1.variable_scope('bn'):
         sess.run(tf.compat.v1.global_variables_initializer())
-        bn_result = sess.run(bnlayer, feed_dict={tf_input: x})
-        grad_result = sess.run(grad, feed_dict={tf_input: x})
+
+        old_var = sess.run(input_variables, feed_dict=f_dict)
+        bn_result = sess.run(bnlayer, feed_dict=f_dict)
+        grad_result = sess.run(grad, feed_dict=f_dict)
+
         updated_gamma = sess.run(input_variables[1] - grad_result[1])
-        updated_beta = sess.run(input_variables[1] - grad_result[2])
+        updated_beta = sess.run(input_variables[2] - grad_result[2])
```

Comment:
How about SGD optimizer here? nntrainer now has the same defaults as tensorflow for SGD. So, IMO, you don't have to worry about learning rate or subtraction, and might be a more robust test.

Reply:
I found out that `tf.keras.optimizer` does not get along with `tf.gradients` as it returns `tf.Tensor` for some and spits out `NotImplementedError` to use `optimizer.get_gradients`, it seems that I need to define some kind of loss. Making a more of an integrated test with loss layer.

While It would be worth testing with loss defined, I think checking the basic local backward pass is more accurate for the purpose. I will just leave the way it is for now and make a patch if I (or others) found a good solution for that.

Reply:
Alright, then add a testcase later with loss.